api/src/main/kotlin/org/jellyfin/apiclient/api/sockets/WebSocketApi.kt

```diff
@@ -0,0 +1,273 @@
+			api.client.ws({
+				url {
+					takeFrom(api.createUrl(
+						pathTemplate = "/socket",
+						queryParameters = mapOf(
+							"api_key" to accessToken,
+							"deviceId" to api.deviceInfo.id
+						)
+					))
+
+					protocol = when (protocol) {
+						URLProtocol.HTTP -> URLProtocol.WS
+						URLProtocol.HTTPS -> URLProtocol.WSS
+
+						// Default to WS
+						else -> URLProtocol.WS
+					}
+				}
+			}) {
+				// Bind to messaging channels
+				joinAll(
+					launch { incoming.read() },
+					launch { outgoing.write() },
+				)
+			}
+		}
+	}
+
+	private fun Frame.readSocketMessage(): IncomingSocketMessage? {
+		require(this is Frame.Text)
+
+		// Read text from frame
+		val text = readText()
+		logger.info("Received message {}", text)
+
+		// Read JSON object from text
+		val json = json.parseToJsonElement(text)
+		require(json is JsonObject)
+
+		// Read id
+		val messageId = json[JSON_PROP_MESSAGE_ID]?.jsonPrimitive?.content
+		require(messageId is String)
+
+		// Read type
+		val type = json[JSON_PROP_MESSAGE_TYPE]?.jsonPrimitive?.content
+		require(type is String)
+
+		// Get serializer for type
+		val dataSerializer = incomingMessageSerializers[type]
+		if (dataSerializer == null) {
+			logger.warn("Message type {} appears to be missing in the incomingMessageSerializers map.", type)
+			return null
+		}
+
+		// Modify JSON to flatten the Data object
+		val modifiedJson = buildJsonObject {
+			put(JSON_PROP_MESSAGE_ID, messageId)
+
+			// Flatten data object or keep it when it's not an object
+			val data = json[JSON_PROP_DATA]
+			if (data is JsonObject) data.entries.forEach { (key, value) -> put(key, value) }
+			else put(JSON_PROP_DATA, data!!)
+		}
+
+		return api.json.decodeFromJsonElement(dataSerializer, modifiedJson) as? IncomingSocketMessage
+	}
+
+	/**
+	 * Publish a message to the server.
+	 */
+	suspend inline fun <reified T : OutgoingSocketMessage> publish(message: T) = publish(message, serializer())
+
+	/**
+	 * Publish a message to the server.
+	 */
+	suspend fun <T : OutgoingSocketMessage> publish(message: T, serializer: KSerializer<T>) {
+		val jsonObject = json.encodeToJsonElement(serializer, message).jsonObject
+		val messageType = serializer.descriptor.serialName
+
+		val text = json.encodeToString(buildJsonObject {
+			put(JSON_PROP_MESSAGE_TYPE, messageType)
+
+			val data = jsonObject[JSON_PROP_DATA]
+			if (data != null) put(JSON_PROP_DATA, data)
+			else putJsonObject(JSON_PROP_DATA) {
+				jsonObject.entries
+					.filterNot { (key, _) -> key == JSON_PROP_MESSAGE_TYPE }
+					.forEach { (key, value) -> put(key, value) }
+			}
+		})
+
+		outgoingMessageChannel.send(text)
+	}
+
+	/**
+	 * Start listening to messages. Calls the [block] for each incoming message until
+	 * [SocketSubscription.cancel] is invoked.
+	 */
+	suspend fun subscribe(block: (IncomingSocketMessage) -> Unit): SocketSubscription {
+		val subscription = SocketSubscription(this, block)
+		subscriptions += subscription
```

Comment:
:eyes: You can do this with LISTS??? That is awesome :rofl:

Reply:
Yeah Kotlin converts all operators to functions on the left-side objects. You can even implement operators on your own classes!